### ISSUE TITLE
Asynchronous resource creation

### ISSUE THREAD

**Issue by kvark** [gfx-rs/gfx#570]
Currently the device trait has all the methods for creating and updating resources. Next-gen APIs as well as DX11 allow creating resources asynchronously:

https://msdn.microsoft.com/en-us/library/windows/desktop/ff476889%28v=vs.85%29.aspx

This is crucial for us to support. The way I see it is having a `Factory` trait that has all these methods. Then, some back-ends will just implement `Factory` for their devices, some will provide `create_factory` methods to support multiple factories operating simultaneously.

**Comment by bvssvni** [comment]
This might solve the problems with glyph caching, where textures needs to be created by demand.

**Comment by kvark** [comment]
@bvssvni This issue is more about DX11/next-gen, since I don't know how this is gonna work with OpenGL. The issue with glyph caching will be addressed in a more straightforward way, I just didn't happen to look into it yet (sorry!).

**Comment by kvark** [comment]
I'm thinking that we could transform `Resources` to fill the factory role:
```rust
trait Device {
   type Resourcer: Resourcer + Mapper;
   type CommandBuffer: CommandBuffer<Self::Resourcer>;

   fn get_capabilities<'a>(&'a self) -> &'a Capabilities;
   fn spawn_resourcer(&self) -> Self::Resourcer;
   fn reset_state(&mut self);
   fn submit(&mut self, buffer: (&Self::CommandBuffer<Self::Resourcer>, &draw::DataBuffer));
}

trait CommandBuffer<R: Resourcer> {
   ...
}

trait Resourcer {
   type X; ...
   fn create_x() ...
   fn delete_x() ...
   fn map_x()...
}
```

**Comment by kvark** [comment]
We could actually go further by splitting different resource types to be handled by different factories:
```rust
trait BufferFactory {
    type Buffer;
    fn create_buffer_raw();
    fn create_buffer<T>();
    fn create_buffer_static_raw();
    fn create_buffer_static<T: Copy>();
}
```
Aside from the architectural (decoupling!) benefit, I can see the following problems with it:

1: `BufferHandle` should be parametrized by what? We don't want users to even think `BufferFactory` exists in the first place.

Possible solution - return tuples, then have typedefs at the top level:
```rust
type BufferHandle<D: Device> = (D::Resourcer::Buffer, BufferInfo);
```
2: How to hide all the traits from the user? Solutions:
  1. perhaps, inheriting `Resources: BufferFactory` allows the user to avoid using `BufferFactory` automatically - we need to verify that.
  2. have all the traits re-exported in the special `mod traits`, so that the user can just do `use gfx::traits::*` and be done with it.

**Comment by kvark** [comment]
The basis (`Factory`) has landed in #632. Changing status to blocked until we have an API backend that actually supports it.

**Comment by kvark** [comment]
D3D11 backend supports it.
